File: apple/Packages/DsmCore/Sources/ContainerNetworkCreation.swift
import Foundation
import Network

/// 创建网络的表单配置；自动 IPv4 和关闭 IPv6 时不提交相应地址字段。
public struct ContainerNetworkCreation: Equatable, Sendable {
    public var name: String
    public var usesManualIPv4 = false
    public var subnet = ""
    public var ipRange = ""
    public var gateway = ""
    public var isIPv6Enabled = false
    public var ipv6Subnet = ""
    public var ipv6Range = ""
    public var ipv6Gateway = ""
    public var disableMasquerade = false

    public init(name: String = "") { self.name = name }

    public enum ValidationIssue: String, Error, Sendable {
        case name = "container.network.validation.name"
        case ipv4 = "container.network.validation.ipv4"
        case ipv6 = "container.network.validation.ipv6"
        case outsideSubnet = "container.network.validation.outsideSubnet"
    }

    public var validationIssue: ValidationIssue? {
        guard let range = name.range(of: "^[a-zA-Z0-9][a-zA-Z0-9_.-]*$", options: .regularExpression),
              range == name.startIndex..<name.endIndex else { return .name }
        if usesManualIPv4 {
            if let issue = Self.validate(subnet: subnet, range: ipRange, gateway: gateway, ipv6: false) { return issue }
        }
        if isIPv6Enabled {
            if let issue = Self.validate(subnet: ipv6Subnet, range: ipv6Range, gateway: ipv6Gateway, ipv6: true) { return issue }
        }
        return nil
    }

    private static func validate(subnet: String, range: String, gateway: String, ipv6: Bool) -> ValidationIssue? {
        let invalid: ValidationIssue = ipv6 ? .ipv6 : .ipv4
        guard let network = cidr(subnet, ipv6: ipv6), let gatewayBytes = address(gateway, ipv6: ipv6) else { return invalid }
        if !ipv6, !(1...223).contains(Int(gatewayBytes[0])) { return invalid }
        guard contains(gatewayBytes, in: network) else { return .outsideSubnet }
        if !range.isEmpty {
            guard let pool = cidr(range, ipv6: ipv6) else { return invalid }
            guard pool.prefix >= network.prefix, contains(pool.bytes, in: network) else { return .outsideSubnet }
        }
        return nil
    }

    private static func address(_ text: String, ipv6: Bool) -> [UInt8]? {
        guard !text.contains("%") else { return nil }
        if ipv6 { return IPv6Address(text).map { Array($0.rawValue) } }
        // 与官方 IPv4 表单一致，不接受省略段或带前导零的十进制段。
        let parts = text.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4, parts.allSatisfy({ part in
            guard let value = Int(part), (0...255).contains(value) else { return false }
            return String(value) == part
        }) else { return nil }
        return IPv4Address(text).map { Array($0.rawValue) }
    }

    private static func cidr(_ text: String, ipv6: Bool) -> (bytes: [UInt8], prefix: Int)? {
        let parts = text.split(separator: "/", omittingEmptySubsequences: false)
        guard parts.count == 2, let prefix = Int(parts[1]), String(prefix) == parts[1],
              (0...(ipv6 ? 128 : 32)).contains(prefix), let bytes = address(String(parts[0]), ipv6: ipv6) else { return nil }
        return (bytes, prefix)
    }

    private static func contains(_ bytes: [UInt8], in network: (bytes: [UInt8], prefix: Int)) -> Bool {
        for index in bytes.indices {
            let bits = min(8, max(0, network.prefix - index * 8))
            let mask: UInt8 = bits == 0 ? 0 : UInt8(255 << (8 - bits) & 255)
            if bytes[index] & mask != network.bytes[index] & mask { return false }
        }
        return true
    }
}

File: windows/src/LanStash.Domain/Containers/ContainerNetworkCreation.cs
using System.Globalization;
using System.Net;
using System.Net.Sockets;

namespace LanStash.Domain;

public enum ContainerNetworkValidationIssue { None, Name, Ipv4, Ipv6, OutsideSubnet }
public sealed record ContainerNetworkCreation(string Name, bool UsesManualIpv4 = false, string Subnet = "", string IpRange = "", string Gateway = "",
    bool IsIpv6Enabled = false, string Ipv6Subnet = "", string Ipv6Range = "", string Ipv6Gateway = "", bool DisableMasquerade = false)
{
    public ContainerNetworkValidationIssue ValidationIssue
    {
        get
        {
            if (string.IsNullOrEmpty(Name) || !char.IsAsciiLetterOrDigit(Name[0]) || Name.Any(c => !char.IsAsciiLetterOrDigit(c) && !"_.-".Contains(c))) return ContainerNetworkValidationIssue.Name;
            if (UsesManualIpv4 && ValidateNetwork(Subnet, IpRange, Gateway, false) is var v4 && v4 != ContainerNetworkValidationIssue.None) return v4;
            if (IsIpv6Enabled && ValidateNetwork(Ipv6Subnet, Ipv6Range, Ipv6Gateway, true) is var v6 && v6 != ContainerNetworkValidationIssue.None) return v6;
            return ContainerNetworkValidationIssue.None;
        }
    }
    public override string ToString() => nameof(ContainerNetworkCreation);
    private static ContainerNetworkValidationIssue ValidateNetwork(string subnet, string range, string gateway, bool ipv6)
    {
        var invalid = ipv6 ? ContainerNetworkValidationIssue.Ipv6 : ContainerNetworkValidationIssue.Ipv4;
        if (range is null) return invalid;
        var network = Cidr(subnet, ipv6); var address = Address(gateway, ipv6);
        if (network is null || address is null || !ipv6 && address[0] is not (>= 1 and <= 223)) return invalid;
        if (!Contains(address, network.Value)) return ContainerNetworkValidationIssue.OutsideSubnet;
        if (!string.IsNullOrEmpty(range))
        {
            var pool = Cidr(range, ipv6);
            if (pool is null) return invalid;
            if (pool.Value.Prefix < network.Value.Prefix || !Contains(pool.Value.Bytes, network.Value)) return ContainerNetworkValidationIssue.OutsideSubnet;
        }
        return ContainerNetworkValidationIssue.None;
    }
    private static byte[]? Address(string text, bool ipv6)
    {
        if (string.IsNullOrEmpty(text) || text.Any(char.IsWhiteSpace) || text.IndexOfAny(['%', '[', ']']) >= 0) return null;
        if (!ipv6)
        {
            var parts = text.Split('.');
            if (parts.Length != 4 || parts.Any(part => !byte.TryParse(part, NumberStyles.None, CultureInfo.InvariantCulture, out var value) || value.ToString(CultureInfo.InvariantCulture) != part)) return null;
        }
        return IPAddress.TryParse(text, out var address) && address.AddressFamily == (ipv6 ? AddressFamily.InterNetworkV6 : AddressFamily.InterNetwork) ? address.GetAddressBytes() : null;
    }
    private static (byte[] Bytes, int Prefix)? Cidr(string text, bool ipv6)
    {
        if (text is null) return null;
        var parts = text.Split('/');
        if (parts.Length != 2 || !int.TryParse(parts[1], NumberStyles.None, CultureInfo.InvariantCulture, out var prefix) ||
            prefix.ToString(CultureInfo.InvariantCulture) != parts[1] || prefix > (ipv6 ? 128 : 32) || Address(parts[0], ipv6) is not { } bytes) return null;
        return (bytes, prefix);
    }
    private static bool Contains(byte[] address, (byte[] Bytes, int Prefix) network)
    {
        for (var index = 0; index < address.Length; index++)
        {
            var bits = Math.Clamp(network.Prefix - index * 8, 0, 8); var mask = bits == 0 ? 0 : (255 << (8 - bits)) & 255;
            if ((address[index] & mask) != (network.Bytes[index] & mask)) return false;
        }
        return true;
    }
    public static bool EquivalentIpv4Cidr(string left, string right)
    {
        var a = Cidr(left, false); var b = Cidr(right, false);
        return a is not null && b is not null && a.Value.Prefix == b.Value.Prefix && Contains(a.Value.Bytes, b.Value);
    }
}
public sealed record ContainerNetworkCreateRequest(Guid ProfileId, ContainerNetworkCreation Configuration, Guid RequestId, bool RiskConfirmed)
{
    public override string ToString() => nameof(ContainerNetworkCreateRequest);
}
public sealed record ContainerNetworkCreationRecovery(string Name)
{
    public override string ToString() => nameof(ContainerNetworkCreationRecovery);
}
